Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,112 @@
+
+-- Multiplication of matrices
+instance Num SBMatrix where
+  (+) (SBMatrix m1) (SBMatrix m2) = SBMatrix (I.unionWith (<>) m1 m2)
+  (*) m1 (SBMatrix m2) = SBMatrix (I.map (m1 `mult`) m2)
+  negate = id
+
+instance Semigroup SBMatrix where
+  (<>) = (*)
+
+size :: SBMatrix -> Int
+size (SBMatrix m) = I.size m
+
+randomSublist :: [Int] -> StdGen -> [Int]
+randomSublist list gen = fmap fst $ filter snd $ zip list (randoms gen)
+
+-- linearSolve :: SBMatrix -> SBVector
+-- linearSolve matrix@(SBMatrix m) = findSolution $ almostZeroMatrix `mult` w
+--   where
+--     -- Rows of z are indexed by the columns of matrix
+--     z = SBVector (S.fromList $ randomSublist (I.keys m) (mkStdGen (fromIntegral (unsafePerformIO getMonotonicTimeNSec))))
+--     randomSequence = generateData matrix z
+--     -- The order of the arguments is exchanged
+--     minPoly = snd (gcdExt randomSequence (SBPolynomial (True : (L.replicate (2 * (size matrix)) False))))
+--     reducedMinPoly = SBPolynomial (L.dropWhileEnd (not . id) (polynomial minPoly))
+--     almostZeroMatrix = evaluate reducedMinPoly matrix
+--     w = SBVector (S.fromList $ randomSublist (I.keys m) (mkStdGen (fromIntegral (unsafePerformIO getMonotonicTimeNSec))))
+--     findSolution :: SBVector -> SBVector
+--     findSolution v
+--       | set (matrix `mult` v) == mempty = v
+--       | otherwise                       = findSolution $ matrix `mult` vList
+
+linear :: SBMatrix -> F2Poly
+linear matrix@(SBMatrix m) = minPoly
+  where
+    z = SBVector (S.fromList $ randomSublist (I.keys m) (mkStdGen (fromIntegral (unsafePerformIO getMonotonicTimeNSec))))
+    randomSequence = generateData matrix z
+    dim = size matrix
+    -- Is this the best way to generate @errorPoly@?
+    errorPoly = toF2Poly $ U.update (U.replicate (2 * dim + 1) (Bit False)) (U.singleton (2 * dim, Bit True))
+    minPoly = berlekampMassey dim errorPoly randomSequence
+
+-- The input is a matrix B and a random vector z
+generateData :: SBMatrix -> SBVector -> F2Poly
+generateData matrix z = toF2Poly $ traceShowId $ U.fromList $ reverse $ map (\v -> Bit (not $ S.null (set (x `mult` v)))) matrixPowers
+  where
+    matrixPowers = L.take (2 * size matrix) $ L.iterate (matrix `mult`) z
+    x = SBMatrix (foldr (\p acc -> I.insert p (SBVector (S.singleton 1)) acc) initialMap randomPrimes)
+    randomPrimes = randomSublist primes (mkStdGen (fromIntegral (unsafePerformIO getMonotonicTimeNSec)))
+    initialMap = I.fromList ([(p, mempty) | p <- primes])
+    -- This should be replaced by factorBase = [nextPrime 2..precPrime b]
+    -- This assumes rows are indexed in the same way as columns are.
+    primes = [1..(size matrix)]
+
+berlekampMassey :: Int -> F2Poly -> F2Poly -> F2Poly
+berlekampMassey dim = go 1 0
+  where
+    -- Is there a better way to implement recursion in this situation?
+    go :: F2Poly -> F2Poly -> F2Poly ->F2Poly -> F2Poly
+    go oneBefore twoBefore a b
```

Comment:
It would be nice to state in comments an invariant relation between `oneBefore`, `twoBefore`, `a` and `b`.